python/gen_requirements.py

```diff
@@ -0,0 +1,391 @@
+      assert isinstance(p, str), f"problems element not a str: {p}"
+      formatted.append('\n'.join(
+        textwrap.wrap(f"{config}: {p}", width=80, initial_indent=" * ", subsequent_indent="   ")))
+
+    return '\n'.join(formatted)
+
+  def __init__(self, config, problems):
+    super(ValidationError, self).__init__(self.format_problems(config, problems))
+    self.problems = problems
+
+
+def validate_or_raise():
+  problems = validate_requirements_by_piece()
+  if problems:
+    raise ValidationError("REQUIREMENTS_BY_PIECE", problems)
+
+  problems = validate_constraints()
+  if problems:
+    raise ValidationError("CONSTRAINTS", problems)
+
+
+def semver_to_requirements(piece, dep, constraint, joined_deps):
+  problems = []
+  min_ver_parts, fixed_index = parse_semver(dep, constraint, problems)
+  assert not problems
+
+  # i is now the index of the version component which must not change.
+  try:
+    fixed_part = int(min_ver_parts[fixed_index])
+  except ValueError:
+    problems.append(
+      f"piece {piece}: dependency {dep} has semver constraint with non-integer leading parts: {ver}")
+    return
+
+  max_ver_parts = (min_ver_parts[:fixed_index] +
+                   [str(int(min_ver_parts[fixed_index]) + 1)] +
+                   ["0" for _ in min_ver_parts[fixed_index + 1:]])
+  joined_deps.append(f'{dep}>={".".join(min_ver_parts)},<{".".join(max_ver_parts)}')
+
+
+def join_requirements():
+  """Validate, then join REQUIRMENTS_BY_PIECE against CONSTRAINTS and return the result.
+
+  Returns
+  -------
+  An OrderedDict containing REQUIREMENTS_BY_PIECE, except any dependency mentioned in CONSTRAINTS
+  is replaced by a setuptools-compatible constraint.
+  """
+  validate_or_raise()
+
+  constraints_map = collections.OrderedDict(
+    [(p.lower(), c) for (p, c) in CONSTRAINTS])
+
+  to_return = collections.OrderedDict()
+  all_deps = set()
+  for piece, deps in REQUIREMENTS_BY_PIECE:
+    joined_deps = []
+    for d in deps:
+      constraint = constraints_map.get(d.lower())
+      if constraint is None:
+        joined_deps.append(d)
+        continue
+
+      if constraint[0] == "^":
+        # Translate semver to requirements:
+        semver_to_requirements(piece, d, constraint, joined_deps)
+      else:
+        joined_deps.append(f"{d}{constraint}")
+
+    if piece != "dev":
+      all_deps.update(joined_deps)
+
+    to_return[piece] = joined_deps
+
+  to_return["all"] = all_deps
+
+  return to_return
+
+
+def parse_args():
+  parser = argparse.ArgumentParser()
+  parser.add_argument("--lint", action="store_true",
+                      help="Just lint dependencies, don't generate anything")
+  return parser.parse_args()
+
+
+def main():
+  args = parse_args()
+  try:
+    joined_deps = join_requirements()
+  except ValidationError as e:
+    print(f"ERROR: invalid requirements configuration in {__file__}:", file=sys.stderr)
+    print(str(e), file=sys.stderr)
+    sys.exit(2)
+
+  if args.lint:
+    sys.exit(0)
```

Comment:
Recommend `return`, not `exit`, since this is a function that could be (and, in fact, is!) invoked by another Python library -- same for the error cases, raise an Exception, don't exit.